metrics/stackdriver_exporter.go

```diff
@@ -63,10 +65,18 @@ func newOpencensusSDExporter(o stackdriver.Options) (view.Exporter, error) {
 // TODO should be properly refactored to be able to inject the getMonitoredResourceFunc function.
 // 	See https://github.com/knative/pkg/issues/608
 func newStackdriverExporter(config *metricsConfig, logger *zap.SugaredLogger) (view.Exporter, error) {
-	gm := gcpMetadataFunc()
+	gm := getGCPMetadata(config)
 	mtf := getMetricTypeFunc(config.stackdriverMetricTypePrefix, config.stackdriverCustomMetricTypePrefix)
-	e, err := newStackdriverExporterFunc(stackdriver.Options{
-		ProjectID:               config.stackdriverProjectID,
+	co, err := getStackdriverExporterClientOptions(&config.stackdriverClientConfig)
+	if err != nil {
+		logger.Warnw("Issue configuring Stackdriver exporter client options, no additional client options will be used: ", zap.Error(err))
+	}
+	// Automatically fall back on Google application default credentials
+	e, err := newOpencensusSDExporter(stackdriver.Options{
+		ProjectID:               gm.project,
```

Comment:
Observation: I don't think you need to merge with the GCE metadata server here, since ProjectID says:

```golang
    // ProjectID is the identifier of the Stackdriver
    // project the user is uploading the stats data to.
    // If not set, this will default to your "Application Default Credentials".
    // For details see: https://developers.google.com/accounts/docs/application-default-credentials.
    //
    // It will be used in the project_id label of a Stackdriver monitored
    // resource if the resource does not inherently belong to a specific
    // project, e.g. on-premise resource like k8s_container or generic_task.
```

This would mean that you could simply drop in the result of fetching the key from the confiiguration map, as empty string would mean "use the GCE default".

Reply:
I believe you are correct. However, I am going to be overly cautious and defer it to a separate PR. Because it would be a deviation from existing behavior and can only be properly tested on GCE, I think it deserves extra validation outside of the scope of this change.